Use original method as HTTP span name

lmolkova opened the pull request:
Fixes #268

## Changes

Uses `http.request.method_original` as a primary source and `http.request.method` as a fallback for HTTP span names.
## Merge requirement checklist

* [ ] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [ ] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [ ] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

trask commented:
does this fit with https://github.com/open-telemetry/opentelemetry-specification/blob/v1.22.0/specification/trace/api.md#span?

> The span name SHOULD be the most general string that identifies a (statistically) interesting *class of Spans*, rather than individual Span instances while still being human-readable.

trask commented on docs/http/http-spans.md:
```diff
@@ -61,13 +61,20 @@ and various HTTP versions like 1.1, 2 and SPDY.
 HTTP client spans have no `http.route` attribute since client-side instrumentation
-is not generally aware of the "route", and therefore HTTP client spans SHOULD use
-`{http.request.method}`.
+is not generally aware of the "route", and therefore HTTP client spans SHOULD be
+`{method}`.
+
+The `{method}` SHOULD be `{http.request.method}` if method represents original method known to the instrumentation.
+In other cases (when `{http.request.method}` is set to `OTHER`), `{method}` SHOULD match `HTTP`.
```
```suggestion
In other cases (when `{http.request.method}` is set to `_OTHER`), `{method}` SHOULD match `HTTP`.
```

Oberon00 replied:
```suggestion
is not generally aware of the "route", and therefore HTTP client spans SHOULD be
` HTTP {method}`.
```

Oberon00 commented on docs/http/http-spans.md:
```diff
@@ -61,13 +61,20 @@ and various HTTP versions like 1.1, 2 and SPDY.
-HTTP server span names SHOULD be `{http.request.method}` if there is no (low-cardinality)
-`http.route` available.
+
+HTTP server span names SHOULD be `{method} {http.route}` if there is a
+(low-cardinality) `http.route` available,
+
+If there is no (low-cardinality) `http.route` available, HTTP server span names
+SHOULD match `{method}`
```
```suggestion
HTTP server span names SHOULD be `HTTP {method} {http.route}` if there is a
(low-cardinality) `http.route` available,

If there is no (low-cardinality) `http.route` available, HTTP server span names
SHOULD match `HTTP {method}`
```

joaopgrassi replied:
Just thinking out loud: Would such change now be considered a breaking change for users? Do people usually rely on the names to build custom alerts or such things? 🤔

mateuszrzeszutek replied:
Doesn't that reverse https://github.com/open-telemetry/opentelemetry-specification/pull/3165 though?

Oberon00 replied:
@mateuszrzeszutek  Yes it does revert that, but not only.  The reason for the HTTP removal IIUC was to keep route-based and method-based span names consistent, which is achieved here by adding the HTTP prefix also for route-based span names.

Oberon00 reviewed (commented):
I liked the suggestion of @trask on the issue to always prepend "HTTP ", even for the route. This will also ensure that HTTP span names don't conflict with other span names from unrelated conventions.

I added that as suggestions to this PR review.

joaopgrassi commented on docs/http/http-spans.md:
```diff
@@ -61,13 +61,20 @@ and various HTTP versions like 1.1, 2 and SPDY.
+is not generally aware of the "route", and therefore HTTP client spans SHOULD be
+`{method}`.
+
+The `{method}` SHOULD be `{http.request.method}` if method represents original method known to the instrumentation.
```
```suggestion
The `{method}` SHOULD be `{http.request.method}` if the method represents the original method known to the instrumentation.
```

trask commented:
> I liked the suggestion of @trask on the issue to always prepend "HTTP ", even for the route. This will also ensure that HTTP span names don't conflict with other span names from unrelated conventions.
> 
> I added that as suggestions to this PR review.

my intention was only to use `HTTP` in place of `_OTHER`, as I think known http verbs are generally descriptive enough for http span names without the addition of `HTTP`

Oberon00 commented:
@trask Sorry, then I misunderstood you.

Oberon00 commented on docs/http/http-spans.md:
```diff
@@ -61,13 +61,20 @@ and various HTTP versions like 1.1, 2 and SPDY.
-HTTP server span names SHOULD be `{http.request.method}` if there is no (low-cardinality)
-`http.route` available.
+
+HTTP server span names SHOULD be `{method} {http.route}` if there is a
```
It is unclear what "method" is at this point, since it's only defined later.

```suggestion
HTTP server span names SHOULD be `{method} {http.route}` (see below for the exact definition of the method placeholder) if there is a
```

Oberon00 commented on docs/http/http-spans.md:
```diff
@@ -61,13 +61,20 @@ and various HTTP versions like 1.1, 2 and SPDY.
+(low-cardinality) `http.route` available,
+
+If there is no (low-cardinality) `http.route` available, HTTP server span names
+SHOULD match `{method}`
```
```suggestion
SHOULD be `{method}` (see below for the exact definition of the method placeholder)
```

Oberon00 commented on docs/http/http-spans.md:
```diff
@@ -61,13 +61,20 @@ and various HTTP versions like 1.1, 2 and SPDY.
+is not generally aware of the "route", and therefore HTTP client spans SHOULD be
+`{method}`.
+
+The `{method}` SHOULD be `{http.request.method}` if the method represents the original method known to the instrumentation.
```
What other options are viable for "method"? If none, I'd recommend MUST

joaopgrassi reviewed (approved):
I'm thinking since this is breaking, if we have a way/process to communicate the community/maintainers/sigs about this, so it's not forgotten? It's probably a general challenge we have here since PRs gets merged but I don't think there's a process where changes such as this are communicated to all SIGs.

Also, I'm not even sure if any instrumentation already implemented the `_OTHER` behavior anyway so might not be that critical.

jsuereth reviewed (approved):
I like the suggestion of always having HTTP as a prefix.

trask commented:
> I like the suggestion of always having HTTP as a prefix.

just confirming, you would like to revisit https://github.com/open-telemetry/opentelemetry-specification/pull/3165?

jsuereth commented:
> just confirming, you would like to revisit https://github.com/open-telemetry/opentelemetry-specification/pull/3165?

Given history there, no.  Only if the idea was new and hand't been considered before.

This is good to merge as-is I think.

Oberon00 commented:
> Given history there, no. Only if the idea was new and hand't been considered before.

The idea of also prepending HTTP to routes for consistency is new IIUC

trask commented:
> The idea of also prepending HTTP to routes for consistency is new IIUC

This is true.

I'm ok revisiting this, but it seems to me to come back to "what is the purpose of span names?", which to my understanding is two things:
* primarily: a logical grouping key for spans
* secondarily: a default display for the span in UIs which don't have deeper knowledge of otel semconv

and I'm not sure that prepending HTTP in front of well-known http methods helps either of these things.

trask commented:
If we did go with prepending `HTTP` to span names, would we also prepend `RPC`, `Database`, `Messaging` to those span names? or would the same reasoning for prepending `HTTP` not apply in those cases?

lmolkova commented:
Agree with @trask that `HTTP` prefix does not bring substantial benefits and if we want to bring it back, we should follow the same patterns in other semconv.

If we need a semconv id and be able to group/visualize based on it, this would arguably go beyond default UX and would be best reported as an instrumentation scope attribute. Adding some sort of semconv-id was discussed around the https://github.com/open-telemetry/oteps/pull/172 and was highly controversial. E.g. AFAIK  elasticsearch reports spans that are both HTTP and DB - how they should name spans/report semconv-id for those?

lmolkova commented:
> Additionally: I think the changes are good, but I wonder if we can get things more "direct/clear" if we split and define proper sections for `server` and `client` HTTP span names. Something like:

I like this idea and I tried it, but I'm not quite happy with the outcome:

The sentence  
> Instrumentation MUST NOT default to using URI path as span name, but MAY provide hooks to allow custom logic to override the default span name.

applies to client and server spans (`http.route` is not really available in many HTTP server instrumentations).
So it leaves us with 1 sentence specific to client and server and much more shared text.

I added an invisible anchor to `{method}`  and references to it, hope it looks better now.

joaopgrassi commented:
> I added an invisible anchor to `{method}` and references to it, hope it looks better now.

Yep look much nicer! Thanks for addressing it and good idea with the invisible anchor 🧠

lmolkova commented on package.json:
```diff
@@ -11,7 +11,7 @@
     "gulp": "^4.0.2",
     "js-yaml": "^4.1.0",
-    "markdown-link-check": "3.10.3",
+    "markdown-link-check": "^3.11.2",
```
the version was pinned due to a bug in the tool, which is fixed now. See https://github.com/open-telemetry/opentelemetry-specification/pull/3327
